mgopersist/persist.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package mgopersist defines cached MongoDB-backed global persistent storage for
+// key-value pairs.
+package mgopersist
+
+import (
+	"sync"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var ErrNotFound = errgo.New("persistent data entry not found")
+
+type entry struct {
+	// value holds the BSON-marshaled value. It will be empty
+	// if the entry was not found when fetched.
+	value []byte
+	// expire holds when the value will expire
+	// from the cache. This will be zero when the
+	// value does not exist.
+	expire time.Time
+}
+
+// Store represents a cached set of key-value pairs.
+type Store struct {
+	cacheLifetime time.Duration
+	mu            sync.RWMutex
+	// entries holds all the cached entries.
+	entries map[string]entry
+}
+
+// Refresh forgets all cached items.
+func (s *Store) Refresh() {
+	s.mu.Lock()
+	s.entries = make(map[string]entry)
+	s.mu.Unlock()
+}
+
+// NewStore returns a Store that will cache items for
+// at most the given time.
+func NewStore(cacheLifetime time.Duration) *Store {
+	return &Store{
+		entries:       make(map[string]entry),
+		cacheLifetime: cacheLifetime,
+	}
+}
+
+// Session associates a Store instance with a mongo session.
+type Session struct {
+	*Store
+	coll *mgo.Collection
+}
+
+// entryDoc holds the document that's stored in MongoDB.
+type entryDoc struct {
+	Key string `bson:"_id"`
+	// Value holds the value. We store it as a raw value
+	// so that it looks nice when looking at the collection directly.
+	Value bson.Raw `bson:"value"`
+}
+
+// Session returns a store session that uses the given
+// collection for storage. Each store entry is stored
+// in a document in the collection.
+func (s *Store) Session(coll *mgo.Collection) *Session {
```

Comment:
I changed the interface to pass the session instead of the collection in Session.

I'd prefer not to call Copy in Session, as we've mgosession to bound the number
of sessions and I think that, even though it's inconvenient, we should stick to
the convention of passing in the session that's created for the request.